analysis/src/abstract_state.rs

```diff
@@ -4,7 +4,46 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-pub trait AbstractState {
-    fn join(self, other: Self) -> Self;
-    // ...
+use rustc_middle::mir;
+use rustc_middle::ty::TyCtxt;
+use std::vec::Vec;
+use crate::AnalysisError;
+use serde::Serialize;
+
+
+pub trait AbstractState<'a, 'tcx: 'a>: Clone + Eq + Sized + Serialize {   // Sized needed for apply_terminator_effect's return type
```

Comment:
Could you add a doc comment with a summary of the properties that should be ensured by the implementation? Like "forall x :: join(bottom(), x) == x".